src/gencost/waterfall.py

```diff
@@ -730,9 +733,60 @@ def get_exa_by_generator(self):
                 )
             )
             .drop(columns=["cems_923_merge", "exa_merge"])
+            .merge(
+                self.get_wage_scale(),
+                on=["report_date", "state"],
+                how="left",
+                validate="m:1",
+            )
+            .fillna({"wage_scale": 1})
+            .assign(
+                hrs_in_yr=lambda x: np.where(x.report_date.dt.is_leap_year, 8784, 8760),
+                gross_cf=lambda x: x.gross_generation_mwh
+                / (x.capacity_mw * x.hrs_in_yr),
+            )
+            .merge(
+                pd.read_parquet(
+                    PACKAGE_PATH / "860_FERC_matching_cost_regressions.parquet.gzip",
+                )[["report_year", "inflator_to_2021"]]
+                .drop_duplicates()
+                .assign(
+                    report_date=lambda x: pd.to_datetime(x["report_year"], format="%Y")
+                ),
+                on=["report_date"],
+                how="left",
+            )
+            .assign(
+                real_pollution_control_costs_per_kw=lambda x: x.pollution_control_costs_per_kw
+                * x.inflator_to_2021
+            )  # pandera caught 3 observations w/null prime movers
+            .query("prime_mover.notnull()")
+            .query("prime_mover in @FOSSIL_PRIME_MOVER_MAP")
         )
 
-        return merged.query('_merge == "all"').drop(columns=["_merge"])
+        # fuel fraction calcs from merge all
+        gross_mwh_cols = merged.filter(like="_gross_mwh").columns
+
+        merged[[c.replace("_gross_mwh", "_fraction") for c in gross_mwh_cols]] = (
+            merged[gross_mwh_cols]
+            .divide(merged[gross_mwh_cols].sum(axis=1), axis=0)
+            .fillna(0.0)
+        )
+
+        core_fuels = ["coal_fraction", "natural_gas_fraction", "petroleum_fraction"]
+
+        out = (
+            merged.assign(
+                minor_fuels_fraction=lambda x: x.filter(like="_fraction").sum(axis=1)
+                - x[core_fuels].sum(axis=1)
+            )
+            .query('_merge == "all"')
+            .drop(columns=["_merge", "hrs_in_yr"])
+        )
+
+        self._dfs = self.core_validation(out, level="generator")
+
+        return self._dfs
```

Comment:
Carefull here, `self._dfs` is a dictionary of cached data frames, assigning something to it directly will delete everything already cached and break other things that depend on what's in those caches. 

`get_cems` is a short example of how to use the cache, you check if the requested data frame is already in `self._dfs` if it isn't you make the df and store it with the correct key in the dictionary, then, outside the if block, you return the requested data frame from the dictionary.

Reply:
oh that's really cool, I tried doing a version of this